docs/modules/ROOT/pages/references/architecture/control-api-design.adoc

```diff
@@ -0,0 +1,181 @@
+
+To circumvent the current limitation of Kubernetes to show filtered lists of cluster-scoped resources based on access rights, the virtual object `Organization` represents an organization.
+
+The dynamic filter works through the list of namespaces which represent an organization and checks if the subject has access to the namespace.
+
+[source,yaml]
+----
+apiVersion: appuio.io/v1
+kind: Organization
+metadata:
+  name: acme-corp
+spec:
+  displayName: Acme Corp.
+----
+
+[source,yaml]
+----
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: org-acme-corp <1>
+  labels:
+    appuio.io/resource-type: organization <2>
+    organization.appuio.io/displayName: Acme Corp. <3>
+----
+<1> Prefixed (`org`) name to circumvent possible name collision (e.g. if a company is called "kube-system")
+<2> Identify resource type, used by the API server to filter for namespace representing organizations
+<3> Reflected in the organization object as `spec.displayName`
+
+Working with organization-resources always happens via the `Organization` object, it's not meant to directly CRUD namespaces which represent an organization.
+
+==== RBAC rules
+
+These are `ClusterRole` which can be bound to a subject by either a `ClusterRoleBinding` or a namespaced `RoleBinding`:
+
+`org-creator`:: Allow for creating organizations - on the `Organization` object (ClusterRoleBinding)
+`org-view`:: Namespace scope RoleBinding to allow view access to an organization
+`org-admin`:: Namespace scope RoleBinding to allow admin access to an organization (update, delete)
+
+=== `Invitation` and `InvitationAccept` (CRD)
+
+An invitation for a user to join an organization.
+
+Definition of being part of an organization:
+
+* At least `org-view` role in organization NS
+* Member of organization group in Keycloak
+
+Process:
+
+. An `Invitation` object is created in the organization namespace for which the invitation is meant for
++
+[source,yaml]
+----
+apiVersion: appuio.io/v1
+kind: Invitation
+metadata:
+  name: invite-1
+  namespace: org-acme-corp
+spec:
+  invitationcode: ju1aefaeKooz <1>
+  role: org-viewer <2>
+----
+<1> Generated invitation code
+<2> Role which this invitation assigns
+. A user who wants to accept this invitation creates a cluster-scoped `InvitationAccept` resource
++
+[source,yaml]
+----
+apiVersion: appuio.io/v1
+kind: InvitationAccept
+metadata:
+  name: invite-1-accept
+spec:
+  invitationcode: ju1aefaeKooz
+status:
+  blubb: bla
+----
+. The invitation controller searches through all namespaces to find an `Invitation` object where `spec.invitationcode` is the same
+.. A clusterrole and a clusterrolebinding is generated to allow access to this `InvitationAccept` object, granting access to the creating user
+.. The user is added to the matching rolebinding (`spec.role` in `Invitation`)
+.. The `Invitation` and the `InvitationAccept` object gets deleted
+.. The clusterrole and a clusterrolebinding is deleted
```

Comment:
Alternative idea:
1. Org admin creates invitation with a spec like this:
```yaml
apiVersion: appuio.io/v1
kind: Invitation
metadata:
  name: invite-joe
  namespace: org-acme-corp
spec:
  user: joe.mustermann
  role: org-viewer
status:
  accepted: false
```
2. Invitation controller creates a Role/Rolebinding that allows `joe.mustermann` to view and edit `invite-joe`
3. `joe.mustermann` edits the invite, sets `accepted: true` (alternatively, accept invitation with dedicated `InvitationAccept` resource that refers which invitation, e.g. `spec.invitationRef=invite-joe`)
4. Kyverno policy verifies that only the user in the spec can edit this invitation and set `accepted: true`
5. After accepting the edit, `Invitation` and Rolebinding gets deleted

This would leverage the Kubernetes/Openshift Authentication so that only the targeted user can accept the invitation.

Reply:
With that the org admin would already have to know the details of the user to invite, and it could only target 1 user. I guess we need to clearly define how the invitation feature should work in general (use-case based) before we can create a design. And it might make sense to have that design in a separate page, focusing on it.